### PR TITLE
[systemd] disable default systemd udev rules for interfaces

### LINKED ISSUE #7364: Interface MAC is overwritten
[sonic_dump_arc-switch1025_20210419_080448.tar.gz](https://github.com/Azure/sonic-buildimage/files/6334381/sonic_dump_arc-switch1025_20210419_080448.tar.gz)

#### Description

Interface MAC is overwriten by systemd udev-rules, specifically /usr/lib/systemd/network/99-default.link:

```
[Match]
OriginalName=*

[Link]
NamePolicy=keep kernel database onboard slot path
AlternativeNamesPolicy=database onboard slot path
MACAddressPolicy=persistent
```

#### Steps to reproduce the issue:
1. Run this script in teamd container (autorestart of teamd container is disabled)

```
#!/bin/bash
while :; do
        pkill -9 teamd;
        ip link del PortChannel100 ;
        rm -rf /var/run/teamd/PortChannel100.pid;
        /usr/bin/teamd -r -t PortChannel100 -c '{"device":"PortChannel100","hwaddr":"7c:fe:90:79:c9:00","runner":{"active":true,"name":"lacp"}}' -L /var/warmboot/teamd/ -g -d ;
        ip link set dev PortChannel100 up ; 
        ip link show PortChannel100;
        ip link show PortChannel100 | grep '7c:fe' || { echo "!!!!Reproduced!!!!"; exit 1; }
done
```

#### Describe the results you received:

```
root@arc-switch1025:/# bash -x ./test.sh 
+ :
+ pkill -9 teamd
+ ip link del PortChannel100
Cannot find device "PortChannel100"
+ rm -rf /var/run/teamd/PortChannel100.pid
+ /usr/bin/teamd -r -t PortChannel100 -c '{"device":"PortChannel100","hwaddr":"7c:fe:90:79:c9:00","runner":{"active":true,"name":"lacp"}}' -L /var/warmboot/teamd/ -g -d
Using team device "PortChannel100".
Using PID file "/var/run/teamd/PortChannel100.pid"
This program is not intended to be run as root.
+ ip link set dev PortChannel100 up
+ ip link show PortChannel100
173: PortChannel100: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN mode DEFAULT group default qlen 1000
    link/ether b2:06:cf:53:a0:b5 brd ff:ff:ff:ff:ff:ff
+ ip link show PortChannel100
+ grep 7c:fe
+ echo '!!!!Reproduced!!!!'
!!!!Reproduced!!!!
+ exit 1

```

#### Describe the results you expected:

Interface MAC should be a router MAC from DEVICE_METADATA.

#### Output of `show version`:

```
SONiC Software Version: SONiC.SONIC.202012.65-5bdbfcf_Internal
Distribution: Debian 10.9
Kernel: 4.19.0-12-2-amd64
Build commit: 5bdbfcfb
Build date: Fri Apr 16 07:25:51 UTC 2021
Built by: sw-r2d2-bot@r-build-sonic-ci02

Platform: x86_64-mlnx_msn2700-r0
HwSKU: ACS-MSN2700
ASIC: mellanox
ASIC Count: 1
Serial Number: MT1822K07815
Uptime: 08:04:30 up  6:12,  2 users,  load average: 9.14, 7.55, 7.41

Docker images:
REPOSITORY                    TAG                                IMAGE ID            SIZE
docker-syncd-mlnx             SONIC.202012.65-5bdbfcf_Internal   7d80f7d0ad45        663MB
docker-syncd-mlnx             latest                             7d80f7d0ad45        663MB
docker-teamd                  SONIC.202012.65-5bdbfcf_Internal   c478c46d217a        409MB
docker-teamd                  latest                             c478c46d217a        409MB
docker-nat                    SONIC.202012.65-5bdbfcf_Internal   43fcee4d38fa        411MB
docker-nat                    latest                             43fcee4d38fa        411MB
docker-platform-monitor       SONIC.202012.65-5bdbfcf_Internal   cd58e93763cd        689MB
docker-platform-monitor       latest                             cd58e93763cd        689MB
docker-orchagent              SONIC.202012.65-5bdbfcf_Internal   1851cb404019        427MB
docker-orchagent              latest                             1851cb404019        427MB
docker-fpm-frr                SONIC.202012.65-5bdbfcf_Internal   fa3aa6d71796        427MB
docker-fpm-frr                latest                             fa3aa6d71796        427MB
docker-sflow                  SONIC.202012.65-5bdbfcf_Internal   68b34c420a09        410MB
docker-sflow                  latest                             68b34c420a09        410MB
docker-snmp                   SONIC.202012.65-5bdbfcf_Internal   7096d4353768        439MB
docker-snmp                   latest                             7096d4353768        439MB
docker-router-advertiser      SONIC.202012.65-5bdbfcf_Internal   629b4c15054f        398MB
docker-router-advertiser      latest                             629b4c15054f        398MB
docker-lldp                   SONIC.202012.65-5bdbfcf_Internal   0ddd14ff3a48        438MB
docker-lldp                   latest                             0ddd14ff3a48        438MB
docker-dhcp-relay             SONIC.202012.65-5bdbfcf_Internal   96fbcc19a582        405MB
docker-dhcp-relay             latest                             96fbcc19a582        405MB
docker-database               SONIC.202012.65-5bdbfcf_Internal   29a22ae57c52        398MB
docker-database               latest                             29a22ae57c52        398MB
docker-sonic-telemetry        SONIC.202012.65-5bdbfcf_Internal   c8b6348a8f19        488MB
docker-sonic-telemetry        latest                             c8b6348a8f19        488MB
docker-sonic-mgmt-framework   SONIC.202012.65-5bdbfcf_Internal   74ca0870d2fa        618MB
docker-sonic-mgmt-framework   latest  
```
#### Output of `show techsupport`:

```
(paste your output here or download and attach the file here )
```

### PR DESCRIPTION
**Pull request by stepanblyschak**
#### Why I did it
Fix #7364 

99-default.link - was always in SONiC, but previous systemd (<247) had an issue and it did not work due to issue systemd/systemd#3374. Now systemd 247 works.

However, such policy overrides teamd provided mac address which causes teamd netdev to use a random mac 
address. Therefore, needs to be disabled.

#### How I did it
Disable default systemd udev rules for interfaces.

#### How to verify it

Execute https://github.com/Azure/sonic-buildimage/issues/7364 reproduction script.

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [x] 202012

### CODE REVIEW

File: files/build_templates/sonic_debian_extension.j2 [hunk 1]
```diff
@@ -388,6 +388,10 @@ sudo cp files/dhcp/90-dhcp6-systcl.conf.j2 $FILESYSTEM_ROOT_USR_SHARE_SONIC_TEMP
 sudo cp $IMAGE_CONFIGS/interfaces/init_interfaces $FILESYSTEM_ROOT/etc/network/interfaces
 sudo mkdir -p $FILESYSTEM_ROOT/etc/network/interfaces.d
 
+# Disable default systemd udev rules for interfaces to prevent
+# systemd from overriding interface's MAC address
+sudo ln -sf /dev/null $FILESYSTEM_ROOT_ETC/systemd/network/99-default.link
```
Comment (lguohan) [inline files/build_templates/sonic_debian_extension.j2:393]: what does the default systemd udev rule doing right now?

where can I find the default systemd udev rule?

Comment (stepanblyschak) [reply files/build_templates/sonic_debian_extension.j2:393]:
```
[Match]
OriginalName=*

[Link]
NamePolicy=keep kernel database onboard slot path
AlternativeNamesPolicy=database onboard slot path
MACAddressPolicy=persistent
```

Assigning a generated persistent MAC address on every interface name matching '*' pattern.

Comment (lguohan) [reply files/build_templates/sonic_debian_extension.j2:393]: why this default rule is causing the issue? do you have a full explanation of the problem?

**Review by lguohan — APPROVED** [review]

**Comment by lguohan** [comment]
1.	I update PR 7369, @Stepan Blyshchak, can you check if it is accurate?
2.	Other than teamd netdev, are you aware of any other netdev is affected?
3.	Instead of disable the 99-default.link, is it the same effect if we set the macpolicy=none?

**Comment by stepanblyschak** [comment]
1. Thanks, it is accurate.
2. I was focued on team devices, but this can happen to phy port netdevs and this was observed by our team
3. Yes, according to what man pages say, although I find it safer to disable it completely to not have to worry about debian upsteam changes in 99-default.link. What do you think?

**Comment by lguohan** [comment]
since /etc/systemd/network override /usr/lib/systemd/network, i think it is better to provide our own policy explicitly. we can provide a file with mac address policy as none and copy to /etc/systemd/network, it would be better since in the future we may want to customize the other policy as well.

**Review by lguohan — APPROVED** [review]

**Pull request merged**